File: frontend/lib/api/tokens.ts
/**
 * API client for token management endpoints.
 *
 * Handles all API token CRUD operations.
 */

import { apiRequest } from "@/lib/api-client";

const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export interface APIToken {
  id: string;
  organization_id: string;
  created_by_user_id: string;
  name: string;
  description?: string;
  token_prefix: string;
  scopes: string[];
  rate_limit: number;
  created_at: string;
  expires_at: string;
  last_used?: string;
  revoked: boolean;
  revoked_at?: string;
  revoked_by_user_id?: string;
}

export interface CreateTokenRequest {
  name: string;
  description?: string;
  scopes: string[];
  rate_limit?: number;
  expires_in_days?: number;
}

export interface CreateTokenResponse {
  success: boolean;
  message: string;
  token: string;
  token_id: string;
  token_prefix: string;
  expires_at: string;
}

export interface ListTokensResponse {
  tokens: APIToken[];
  total: number;
  page: number;
  page_size: number;
}

export interface RotateTokenResponse {
  success: boolean;
  message: string;
  new_token: string;
  token_id: string;
  token_prefix: string;
  old_token_expires_at: string;
  new_token_expires_at: string;
}

export interface RevokeTokenResponse {
  success: boolean;
  message: string;
  token_id: string;
  revoked_at: string;
}

/**
 * Get authentication headers with access token.
 */
async function getAuthHeaders(): Promise<HeadersInit> {
  // Get session from NextAuth
  const { getSession } = await import("next-auth/react");
  const session = await getSession();

  if (!session?.accessToken) {
    throw new Error("Not authenticated");
  }

  return {
    "Content-Type": "application/json",
    Authorization: `Bearer ${session.accessToken}`,
  };
}

/**
 * List all API tokens for the organization.
 */
export async function listTokens(params: {
  include_revoked?: boolean;
  page?: number;
  page_size?: number;
}): Promise<ListTokensResponse> {
  const queryParams = new URLSearchParams();
  if (params.include_revoked !== undefined) {
    queryParams.append("include_revoked", params.include_revoked.toString());
  }
  if (params.page) queryParams.append("page", params.page.toString());
  if (params.page_size)
    queryParams.append("page_size", params.page_size.toString());

  return apiRequest<ListTokensResponse>(
    `/v1/tokens?${queryParams.toString()}`
  );
}

/**
 * Create new API token.
 */
export async function createToken(
  data: CreateTokenRequest
): Promise<CreateTokenResponse> {
  return apiRequest<CreateTokenResponse>(`/v1/tokens`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/**
 * Get API token by ID.
 */
export async function getToken(tokenId: string): Promise<APIToken> {
  return apiRequest<APIToken>(`/v1/tokens/${tokenId}`);
}

/**
 * Update API token metadata.
 */
export async function updateToken(
  tokenId: string,
  data: {
    name?: string;
    description?: string;
    scopes?: string[];
  }
): Promise<APIToken> {
  return apiRequest<APIToken>(`/v1/tokens/${tokenId}`, {
    method: "PATCH",
    body: JSON.stringify(data),
  });
}

/**
 * Rotate API token secret.
 */
export async function rotateToken(
  tokenId: string,
  gracePeriodHours: number = 24
): Promise<RotateTokenResponse> {
  return apiRequest<RotateTokenResponse>(`/v1/tokens/${tokenId}/rotate`, {
    method: "POST",
    body: JSON.stringify({ grace_period_hours: gracePeriodHours }),
  });
}

/**
 * Revoke API token (immediate, irreversible).
 */
export async function revokeToken(
  tokenId: string
): Promise<RevokeTokenResponse> {
  return apiRequest<RevokeTokenResponse>(`/v1/tokens/${tokenId}/revoke`, {
    method: "POST",
  });
}

/**
 * Delete API token (permanent).
 */
export async function deleteToken(tokenId: string): Promise<void> {
  return apiRequest<void>(`/v1/tokens/${tokenId}`, {
    method: "DELETE",
  });
}

File: src/api/models/tokens.py
"""
API Token management models (Phase 5 - Web UI).

Pydantic models for API token CRUD operations:
- Token creation and management
- Token rotation and revocation
- Scope-based access control
"""

from pydantic import BaseModel, Field, field_validator
from typing import Optional, Literal
from datetime import datetime


# ========== Request Models ==========


class CreateTokenRequest(BaseModel):
    """
    API token creation request.

    Example:
        {
            "name": "CI/CD Pipeline Token",
            "description": "GitHub Actions for main repo",
            "scopes": ["scan:write", "reference:read"],
            "rate_limit": 1000,
            "expires_in_days": 365
        }
    """
    name: str = Field(..., min_length=3, max_length=100, description="Token name (for identification)")
    description: Optional[str] = Field(None, max_length=500, description="Optional description")
    scopes: list[str] = Field(..., min_items=1, description="Token scopes (e.g., 'scan:write', 'reference:read')")
    rate_limit: Optional[int] = Field(1000, ge=100, le=10000, description="Requests per hour limit")
    expires_in_days: Optional[int] = Field(365, ge=1, le=3650, description="Token expiry in days (default: 365, max: 3650/10 years)")

    @field_validator("scopes")
    @classmethod
    def validate_scopes(cls, v: list[str]) -> list[str]:
        """Validate scopes against allowed values."""
        allowed_scopes = {
            "scan:write",      # Upload SBOM/SARIF
            "scan:read",       # Read scan results
            "reference:read",  # Query reference data (CVEs, NIST, etc.)
            "vex:generate",    # Generate VEX documents
            "admin:all",       # Full admin access (organization owners only)
        }

        for scope in v:
            if scope not in allowed_scopes:
                raise ValueError(f"Invalid scope: {scope}. Allowed: {', '.join(allowed_scopes)}")

        return v


class UpdateTokenRequest(BaseModel):
    """
    API token update request (name, description, scopes).

    Example:
        {
            "name": "Updated Token Name",
            "description": "Updated description",
            "scopes": ["scan:write", "reference:read", "vex:generate"]
        }
    """
    name: Optional[str] = Field(None, min_length=3, max_length=100, description="New token name")
    description: Optional[str] = Field(None, max_length=500, description="New description")
    scopes: Optional[list[str]] = Field(None, min_items=1, description="New scopes")

    @field_validator("scopes")
    @classmethod
    def validate_scopes(cls, v: Optional[list[str]]) -> Optional[list[str]]:
        """Validate scopes against allowed values."""
        if v is None:
            return v

        allowed_scopes = {
            "scan:write",
            "scan:read",
            "reference:read",
            "vex:generate",
            "admin:all",
        }

        for scope in v:
            if scope not in allowed_scopes:
                raise ValueError(f"Invalid scope: {scope}. Allowed: {', '.join(allowed_scopes)}")

        return v


class RotateTokenRequest(BaseModel):
    """
    Token rotation request (generates new secret, old valid for 24h).

    Example:
        {
            "grace_period_hours": 24
        }
    """
    grace_period_hours: int = Field(24, ge=1, le=168, description="Hours old token remains valid (1-168h/7 days)")


# ========== Response Models ==========


class APIToken(BaseModel):
    """
    API token details (secret NOT included after creation).

    Example:
        {
            "id": "token_abc123",
            "organization_id": "org_xyz789",
            "created_by_user_id": "users/user123",
            "name": "CI/CD Pipeline Token",
            "description": "GitHub Actions for main repo",
            "token_prefix": "complira_tk_abc123",
            "scopes": ["scan:write", "reference:read"],
            "rate_limit": 1000,
            "created_at": "2024-01-15T10:00:00Z",
            "expires_at": "2025-01-15T10:00:00Z",
            "last_used": "2024-01-15T14:28:00Z",
            "revoked": false
        }
    """
    id: str = Field(..., description="Token document key")
    organization_id: str = Field(..., description="Organization ID")
    created_by_user_id: str = Field(..., description="User who created the token")
    name: str = Field(..., description="Token name")
    description: Optional[str] = Field(None, description="Token description")
    token_prefix: str = Field(..., description="Token prefix (e.g., 'complira_tk_abc123...')")
    scopes: list[str] = Field(..., description="Token scopes")
    rate_limit: int = Field(..., description="Requests per hour")
    created_at: str = Field(..., description="Creation timestamp (ISO 8601)")
    expires_at: str = Field(..., description="Expiration timestamp (ISO 8601)")
    last_used: Optional[str] = Field(None, description="Last use timestamp (ISO 8601)")
    revoked: bool = Field(..., description="Whether token is revoked")
    revoked_at: Optional[str] = Field(None, description="Revocation timestamp (ISO 8601)")
    revoked_by_user_id: Optional[str] = Field(None, description="User who revoked the token")


class CreateTokenResponse(BaseModel):
    """
    Token creation response (includes full secret - SHOWN ONCE).

    Example:
        {
            "success": true,
            "message": "Token created successfully. Save this token - you won't see it again!",
            "token": "complira_tk_abc123xyz789defghijklmnopqrstuvwxyz0123456789ABCDEFGHIJK",
            "token_id": "token_abc123",
            "token_prefix": "complira_tk_abc123",
            "expires_at": "2025-01-15T10:00:00Z"
        }
    """
    success: bool = Field(..., description="Whether creation succeeded")
    message: str = Field(..., description="Human-readable message")
    token: str = Field(..., description="Full token secret (ONLY shown once)")
    token_id: str = Field(..., description="Token document key")
    token_prefix: str = Field(..., description="Token prefix for UI identification")
    expires_at: str = Field(..., description="Expiration timestamp (ISO 8601)")


class RotateTokenResponse(BaseModel):
    """
    Token rotation response (new secret + grace period info).

    Example:
        {
            "success": true,
            "message": "Token rotated successfully. Old token expires in 24 hours.",
            "new_token": "complira_tk_xyz789abc456newtoken...",
            "token_id": "token_abc123",
            "token_prefix": "complira_tk_xyz789",
            "old_token_expires_at": "2024-01-16T10:00:00Z",
            "new_token_expires_at": "2025-01-15T10:00:00Z"
        }
    """
    success: bool = Field(..., description="Whether rotation succeeded")
    message: str = Field(..., description="Human-readable message")
    new_token: str = Field(..., description="New token secret (ONLY shown once)")
    token_id: str = Field(..., description="Token document key")
    token_prefix: str = Field(..., description="New token prefix")
    old_token_expires_at: str = Field(..., description="When old token becomes invalid (ISO 8601)")
    new_token_expires_at: str = Field(..., description="When new token expires (ISO 8601)")


class RevokeTokenResponse(BaseModel):
    """
    Token revocation response.

    Example:
        {
            "success": true,
            "message": "Token revoked successfully",
            "token_id": "token_abc123",
            "revoked_at": "2024-01-15T15:00:00Z"
        }
    """
    success: bool = Field(..., description="Whether revocation succeeded")
    message: str = Field(..., description="Human-readable message")
    token_id: str = Field(..., description="Token document key")
    revoked_at: str = Field(..., description="Revocation timestamp (ISO 8601)")


class ListTokensResponse(BaseModel):
    """
    List tokens response (paginated).

    Example:
        {
            "tokens": [
                {
                    "id": "token_abc123",
                    "name": "CI/CD Pipeline Token",
                    "token_prefix": "complira_tk_abc123",
                    "scopes": ["scan:write"],
                    "last_used": "2024-01-15T14:28:00Z",
                    "expires_at": "2025-01-15T10:00:00Z",
                    "revoked": false
                }
            ],
            "total": 3,
            "page": 1,
            "page_size": 20
        }
    """
    tokens: list[APIToken] = Field(..., description="List of tokens")
    total: int = Field(..., description="Total tokens (across all pages)")
    page: int = Field(..., description="Current page number")
    page_size: int = Field(..., description="Items per page")
